File: src/core/identityKey.js
// Long-term identity signing key (ECDSA over P-256).
//
// Every device keeps one per-install ECDSA key pair that lives in IndexedDB
// with the private half non-extractable. We use it to sign the wireHello
// handshake payload so a peer can cryptographically verify that the DH pubkey
// they just received actually came from the holder of this identity — that
// is how MITM on the handshake is prevented.
//
// The public SPKI bytes are the stable identity. Their SHA-256 hash is
// exposed to the UI as a short hex fingerprint; users can compare fingerprints
// out-of-band (QR, voice) to detect an attacker who substitutes keys on the
// very first handshake (TOFU bootstrap).

import { base64ToBytes, bytesToBase64 } from './base64.js';
import { openDatabase } from './db.js';

const KEY_ID = 'identity-signing-v1';
const X3DH_KEY_ID = 'identity-x3dh-v1';

// Cache in-memory so we don't round-trip IDB on every sign call.
let cachedKeyPair = null;
let cachedPubSpki = null;
let cachedFingerprint = null;
let cachedX3dhKeyPair = null;
let cachedX3dhPubSpki = null;
let cachedX3dhBinding = null;

async function readKeyPairFromDb() {
  const db = await openDatabase();
  const row = await db.get('keys', KEY_ID);
  if (!row || !row.privateKey || !row.publicKey) return null;
  // Sanity: verify both halves decoded as CryptoKey (IDB structured-clones
  // them). If the browser downgraded to raw JWK, we can't use it.
  if (typeof row.privateKey !== 'object' || typeof row.publicKey !== 'object') {
    return null;
  }
  return { privateKey: row.privateKey, publicKey: row.publicKey };
}

async function writeKeyPairToDb(keyPair) {
  const db = await openDatabase();
  await db.put('keys', {
    id: KEY_ID,
    privateKey: keyPair.privateKey,
    publicKey: keyPair.publicKey,
    createdAt: Date.now()
  });
}

async function generateNewPair() {
  // Non-extractable private key — the DB can structured-clone it but the
  // app itself cannot export the raw bytes, so a future XSS cannot exfiltrate
  // the private key via JS introspection.
  return crypto.subtle.generateKey(
    { name: 'ECDSA', namedCurve: 'P-256' },
    false,
    ['sign', 'verify']
  );
}

/**
 * Return the cached identity signing key pair, creating+persisting one on
 * first call.
 */
export async function getOrCreateSigningKey() {
  if (cachedKeyPair) return cachedKeyPair;
  const existing = await readKeyPairFromDb();
  if (existing) {
    cachedKeyPair = existing;
    return existing;
  }
  const pair = await generateNewPair();
  await writeKeyPairToDb(pair);
  cachedKeyPair = pair;
  // Invalidate derived caches.
  cachedPubSpki = null;
  cachedFingerprint = null;
  return pair;
}

/**
 * SPKI bytes of the local identity public key. The public half is always
 * exportable regardless of the non-extractable flag on the private half.
 */
export async function exportIdentityPubSpki() {
  if (cachedPubSpki) return cachedPubSpki;
  const { publicKey } = await getOrCreateSigningKey();
  const buf = await crypto.subtle.exportKey('spki', publicKey);
  cachedPubSpki = new Uint8Array(buf);
  return cachedPubSpki;
}

/**
 * Compute the SHA-256 fingerprint of a raw SPKI byte array, lowercase hex.
 * Pure function — safe to call on remote pubkeys too.
 */
export async function computeFingerprint(spkiBytes) {
  const buf = await crypto.subtle.digest('SHA-256', spkiBytes);
  const bytes = new Uint8Array(buf);
  let out = '';
  for (let i = 0; i < bytes.length; i++) {
    out += bytes[i].toString(16).padStart(2, '0');
  }
  return out;
}

/**
 * Human-readable local identity fingerprint (64 hex chars). Cached for UI.
 */
export async function getLocalIdentityFingerprint() {
  if (cachedFingerprint) return cachedFingerprint;
  const spki = await exportIdentityPubSpki();
  cachedFingerprint = await computeFingerprint(spki);
  return cachedFingerprint;
}

/**
 * 16-hex-char short fingerprint, for compact display / peer comparison.
 */
export function shortFingerprint(fingerprint) {
  const s = String(fingerprint || '');
  return s.slice(0, 16);
}

/**
 * Sign a byte array with the local identity key. SHA-256 is the digest used
 * by ECDSA — this matches the hash parameter in verifySignature below.
 */
export async function signBytes(dataBytes) {
  const { privateKey } = await getOrCreateSigningKey();
  const sig = await crypto.subtle.sign(
    { name: 'ECDSA', hash: 'SHA-256' },
    privateKey,
    dataBytes
  );
  return new Uint8Array(sig);
}

/**
 * Verify an ECDSA-P256-SHA256 signature against a remote SPKI pubkey.
 * Returns a boolean (never throws for bad inputs — callers decide how to
 * handle verification failure).
 */
export async function verifyWithRemoteSpki(remoteSpkiBytes, dataBytes, sigBytes) {
  if (!remoteSpkiBytes || !dataBytes || !sigBytes) return false;
  try {
    const remoteKey = await crypto.subtle.importKey(
      'spki',
      remoteSpkiBytes,
      { name: 'ECDSA', namedCurve: 'P-256' },
      false,
      ['verify']
    );
    return await crypto.subtle.verify(
      { name: 'ECDSA', hash: 'SHA-256' },
      remoteKey,
      sigBytes,
      dataBytes
    );
  } catch (_) {
    return false;
  }
}

/**
 * Build the canonical byte string that both sides sign. Ordering matters:
 * the same function must produce the same bytes on both peers for the
 * signature to verify. Format:
 *
 *   "orbits-wire-v3\n" || senderPeerId || "\n" || receiverPeerId || "\n" ||
 *   base64(senderDhSpki) || "\n" || base64(senderIdSpki)
 *
 * Binding receiverPeerId blocks "take handshake from chat A, replay to chat
 * B" attacks; binding senderIdSpki blocks a (contrived) attack where two
 * identity keys share a fingerprint prefix.
 */
export function buildSignedHelloBlob({
  senderPeerId,
  receiverPeerId,
  senderDhSpki,
  senderIdSpki,
  x3dhExtras = null
}) {
  // v4 binds X3DH bootstrap material (sender's ECDH identity pub, ephemeral
  // pub, and the ids of bob's prekeys alice consumed) into the hello
  // signature so a MITM can't silently swap any of them. v3 hellos keep the
  // original 5-line blob to stay bit-compatible with already-shipped peers.
  const parts = [
    x3dhExtras ? 'orbits-wire-v4' : 'orbits-wire-v3',
    String(senderPeerId || ''),
    String(receiverPeerId || ''),
    bytesToBase64(senderDhSpki),
    bytesToBase64(senderIdSpki)
  ];
  if (x3dhExtras) {
    parts.push(bytesToBase64(x3dhExtras.x3dhIkSpki));
    parts.push(bytesToBase64(x3dhExtras.ekSpki));
    parts.push(String(x3dhExtras.spkId || ''));
    parts.push(String(x3dhExtras.opkId || ''));
  }
  return new TextEncoder().encode(parts.join('\n'));
}

// ─── X3DH long-term ECDH identity ────────────────────────────────
//
// X3DH's first DH is DH(IK_a, SPK_b) — both sides need an ECDH-capable
// long-term identity. Our `identity-signing-v1` is ECDSA, so we keep a
// second long-term key specifically for X3DH. It's bound to the ECDSA
// identity by a signature: the ECDSA key signs the ECDH public SPKI bytes,
// so a verifier who already trusts the ECDSA identity (via fingerprint /
// TOFU pin) can extend that trust to the ECDH half without a second
// fingerprint comparison.

const X3DH_BINDING_PREFIX = new TextEncoder().encode('orbits-x3dh-ik-v1\n');

function buildX3dhBindingBlob(x3dhPubSpki) {
  const out = new Uint8Array(X3DH_BINDING_PREFIX.length + x3dhPubSpki.length);
  out.set(X3DH_BINDING_PREFIX, 0);
  out.set(x3dhPubSpki, X3DH_BINDING_PREFIX.length);
  return out;
}

async function readX3dhPairFromDb() {
  const db = await openDatabase();
  const row = await db.get('keys', X3DH_KEY_ID);
  if (!row || !row.privateKey || !row.publicKey || !row.bindingSig) return null;
  if (typeof row.privateKey !== 'object' || typeof row.publicKey !== 'object') return null;
  return { privateKey: row.privateKey, publicKey: row.publicKey, bindingSig: row.bindingSig };
}

async function generateX3dhPair() {
  return crypto.subtle.generateKey(
    { name: 'ECDH', namedCurve: 'P-256' },
    false,
    ['deriveBits']
  );
}

/**
 * Return the cached long-term X3DH ECDH pair, creating+persisting one on
 * first call. The freshly-minted pub is signed with the ECDSA identity so
 * remote peers can bind the two halves together.
 */
export async function getOrCreateX3DHIdentity() {
  if (cachedX3dhKeyPair && cachedX3dhBinding) {
    return { ...cachedX3dhKeyPair, bindingSig: cachedX3dhBinding };
  }
  const existing = await readX3dhPairFromDb();
  if (existing) {
    cachedX3dhKeyPair = { privateKey: existing.privateKey, publicKey: existing.publicKey };
    cachedX3dhBinding = existing.bindingSig;
    return existing;
  }

  const pair = await generateX3dhPair();
  const pubBuf = await crypto.subtle.exportKey('spki', pair.publicKey);
  const pubSpki = new Uint8Array(pubBuf);
  const bindingSig = await signBytes(buildX3dhBindingBlob(pubSpki));

  const db = await openDatabase();
  await db.put('keys', {
    id: X3DH_KEY_ID,
    privateKey: pair.privateKey,
    publicKey: pair.publicKey,
    bindingSig,
    createdAt: Date.now()
  });

  cachedX3dhKeyPair = { privateKey: pair.privateKey, publicKey: pair.publicKey };
  cachedX3dhBinding = bindingSig;
  cachedX3dhPubSpki = pubSpki;
  return { privateKey: pair.privateKey, publicKey: pair.publicKey, bindingSig };
}

/** SPKI bytes of the local X3DH ECDH public key. */
export async function exportX3DHIdentityPubSpki() {
  if (cachedX3dhPubSpki) return cachedX3dhPubSpki;
  const { publicKey } = await getOrCreateX3DHIdentity();
  const buf = await crypto.subtle.exportKey('spki', publicKey);
  cachedX3dhPubSpki = new Uint8Array(buf);
  return cachedX3dhPubSpki;
}

/**
 * Verify that an X3DH public SPKI was endorsed by a given ECDSA identity.
 * Callers should first verify the ECDSA identity itself (fingerprint / TOFU
 * pin), then chain through this to trust the ECDH half.
 */
export async function verifyX3DHBinding(identitySpki, x3dhPubSpki, bindingSig) {
  return verifyWithRemoteSpki(identitySpki, buildX3dhBindingBlob(x3dhPubSpki), bindingSig);
}

/** For tests: reset the in-memory cache so a fresh read from IDB happens. */
export function __resetCacheForTests() {
  cachedKeyPair = null;
  cachedPubSpki = null;
  cachedFingerprint = null;
  cachedX3dhKeyPair = null;
  cachedX3dhPubSpki = null;
  cachedX3dhBinding = null;
}

export { bytesToBase64, base64ToBytes };

File: lib/lib/core/identity_key.dart
// Port of src/core/identityKey.js.
//
// Every install keeps two long-term key pairs:
//
//   • ECDSA P-256 signing pair (`identity-signing-v1`) — the stable identity.
//     Its public SPKI is what peers pin via TOFU. It signs handshake blobs
//     (wireHello v3/v4), prekeys, and the X3DH ECDH binding.
//
//   • ECDH P-256 key agreement pair (`identity-x3dh-v1`) — used in DH1/DH2
//     of X3DH. Bound to the ECDSA identity by a signature over its public
//     SPKI, so TOFU on the ECDSA identity transitively trusts the ECDH half
//     without a second fingerprint comparison.
//
// Private keys are held in memory after first load; persistence is delegated
// to [KeyStore] (backed by in-memory for now, swap to Isar/Hive later — the
// JS side uses IndexedDB with non-extractable CryptoKey structured clones,
// which has no Flutter equivalent; we extract raw private-key bytes and
// re-import, relying on the vault KEK + OS secure storage to protect them).

import 'dart:convert';
import 'dart:typed_data';

import 'package:cryptography/cryptography.dart';

import 'base64_helpers.dart';
import 'key_store.dart';
import 'spki_codec.dart';

const String _signingKeyId = 'identity-signing-v1';
const String _x3dhKeyId = 'identity-x3dh-v1';
const String _keysTable = 'keys';
const String _x3dhBindingLabel = 'orbits-x3dh-ik-v1\n';

final _ecdsa = Ecdsa.p256(Sha256());
final _ecdh = Ecdh.p256(length: 32);

EcKeyPair? _cachedSigningKeyPair;
Uint8List? _cachedSigningPubSpki;
String? _cachedFingerprint;

EcKeyPair? _cachedX3dhKeyPair;
Uint8List? _cachedX3dhPubSpki;
Uint8List? _cachedX3dhBindingSig;

// ─────────────────────────────────────────────────────────────
// Generic helpers (private-key round-trip + SPKI export)
// ─────────────────────────────────────────────────────────────

Future<Uint8List> _exportPubSpki(EcKeyPair keyPair) async {
  final pub = await keyPair.extractPublicKey();
  if (pub is! EcPublicKey) {
    throw StateError(
      'identity: expected EcPublicKey, got ${pub.runtimeType}',
    );
  }
  return buildP256Spki(x: pub.x, y: pub.y);
}

Future<Map<String, Object?>> _serializeKeyPair(
  EcKeyPair keyPair, {
  required Uint8List pubSpki,
}) async {
  // cryptography 2.9 removed the `extractPrivateKeyBytes` shortcut for
  // EcKeyPair — `extract()` returns an EcKeyPairData whose `d` is the raw
  // scalar. Same bytes, extra await hop.
  final priv = (await keyPair.extract()).d;
  return {
    'privBytes': Uint8List.fromList(priv),
    'pubSpki': pubSpki,
  };
}

/// Rehydrate a P-256 keypair from a stored row. [algo] controls the returned
/// key type (ECDSA vs ECDH) — both live at the same raw 32-byte secret, but
/// the `cryptography` package keys them apart by KeyPairType on the public
/// half.
Future<EcKeyPair> _importKeyPair(
  Map<String, Object?> row, {
  required bool ecdsa,
}) async {
  final priv = row['privBytes'];
  final pubSpkiRaw = row['pubSpki'];
  if (priv is! List<int> || pubSpkiRaw is! List<int>) {
    throw StateError('identity: stored row is missing key bytes');
  }
  final point = parseP256Spki(pubSpkiRaw);
  // EcKeyPairData needs the scalar (`d`) and both public affine coords
  // separately — unlike the old SimpleKeyPairData shape which bundled the
  // public key object. Same semantic, different constructor.
  return EcKeyPairData(
    d: priv,
    x: point.x,
    y: point.y,
    type: KeyPairType.p256,
  );
}

// ─────────────────────────────────────────────────────────────
// ECDSA signing identity
// ─────────────────────────────────────────────────────────────

/// Return the cached identity signing key pair, creating+persisting one on
/// first call. Mirrors `getOrCreateSigningKey`.
Future<EcKeyPair> getOrCreateSigningKey() async {
  if (_cachedSigningKeyPair != null) return _cachedSigningKeyPair!;

  final existing = await keyStore().get(_keysTable, _signingKeyId);
  if (existing != null) {
    _cachedSigningKeyPair = await _importKeyPair(existing, ecdsa: true);
    _cachedSigningPubSpki = Uint8List.fromList(existing['pubSpki'] as List<int>);
    return _cachedSigningKeyPair!;
  }

  final pair = await _ecdsa.newKeyPair();
  final pubSpki = await _exportPubSpki(pair);
  final row = await _serializeKeyPair(pair, pubSpki: pubSpki);
  await keyStore().put(_keysTable, {
    'id': _signingKeyId,
    ...row,
    'createdAt': DateTime.now().millisecondsSinceEpoch,
  });
  _cachedSigningKeyPair = pair;
  _cachedSigningPubSpki = pubSpki;
  _cachedFingerprint = null;
  return pair;
}

/// SPKI bytes of the local identity public key.
Future<Uint8List> exportIdentityPubSpki() async {
  if (_cachedSigningPubSpki != null) return _cachedSigningPubSpki!;
  final pair = await getOrCreateSigningKey();
  final spki = await _exportPubSpki(pair);
  _cachedSigningPubSpki = spki;
  return spki;
}

/// SHA-256 fingerprint (lowercase hex) over raw SPKI bytes.
Future<String> computeFingerprint(List<int> spkiBytes) async {
  final hash = await Sha256().hash(spkiBytes);
  return hash.bytes.map((b) => b.toRadixString(16).padLeft(2, '0')).join();
}

/// Local identity fingerprint, cached.
Future<String> getLocalIdentityFingerprint() async {
  if (_cachedFingerprint != null) return _cachedFingerprint!;
  final spki = await exportIdentityPubSpki();
  _cachedFingerprint = await computeFingerprint(spki);
  return _cachedFingerprint!;
}

/// 16-hex-char short fingerprint for compact display.
String shortFingerprint(String fingerprint) =>
    fingerprint.length <= 16 ? fingerprint : fingerprint.substring(0, 16);

/// Sign [dataBytes] with the local ECDSA identity. Returns a raw 64-byte
/// R||S signature (IEEE P1363) matching Web Crypto's output.
Future<Uint8List> signBytes(List<int> dataBytes) async {
  final pair = await getOrCreateSigningKey();
  final sig = await _ecdsa.sign(dataBytes, keyPair: pair);
  return Uint8List.fromList(sig.bytes);
}

/// Verify an ECDSA-P256/SHA-256 signature against a remote SPKI pubkey. Never
/// throws for bad inputs — returns false so callers can treat them uniformly.
Future<bool> verifyWithRemoteSpki(
  List<int>? remoteSpkiBytes,
  List<int>? dataBytes,
  List<int>? sigBytes,
) async {
  if (remoteSpkiBytes == null || dataBytes == null || sigBytes == null) {
    return false;
  }
  try {
    final point = parseP256Spki(remoteSpkiBytes);
    final pub = EcPublicKey(
      x: point.x,
      y: point.y,
      type: KeyPairType.p256,
    );
    return await _ecdsa.verify(
      dataBytes,
      signature: Signature(sigBytes, publicKey: pub),
    );
  } catch (_) {
    return false;
  }
}

/// Canonical signed-hello blob. Both sides must generate byte-identical
/// output or the signature won't verify — treat this function as a frozen
/// wire format. Mirrors `buildSignedHelloBlob` in identityKey.js.
///
/// v3: 5 lines. v4 (when [x3dhExtras] is non-null): adds 4 extra lines so
/// the X3DH bootstrap fields are bound into the signature too.
Uint8List buildSignedHelloBlob({
  required String senderPeerId,
  required String receiverPeerId,
  required List<int> senderDhSpki,
  required List<int> senderIdSpki,
  X3dhHelloExtras? x3dhExtras,
}) {
  final parts = <String>[
    x3dhExtras != null ? 'orbits-wire-v4' : 'orbits-wire-v3',
    senderPeerId,
    receiverPeerId,
    bytesToBase64(Uint8List.fromList(senderDhSpki)),
    bytesToBase64(Uint8List.fromList(senderIdSpki)),
  ];
  if (x3dhExtras != null) {
    parts.add(bytesToBase64(Uint8List.fromList(x3dhExtras.x3dhIkSpki)));
    parts.add(bytesToBase64(Uint8List.fromList(x3dhExtras.ekSpki)));
    parts.add(x3dhExtras.spkId);
    parts.add(x3dhExtras.opkId ?? '');
  }
  return Uint8List.fromList(utf8.encode(parts.join('\n')));
}

/// X3DH fields that ride inside the signed hello when v4 bootstrap is used.
class X3dhHelloExtras {
  const X3dhHelloExtras({
    required this.x3dhIkSpki,
    required this.ekSpki,
    required this.spkId,
    this.opkId,
  });
  final List<int> x3dhIkSpki;
  final List<int> ekSpki;
  final String spkId;
  final String? opkId;
}

// ─────────────────────────────────────────────────────────────
// X3DH long-term ECDH identity
// ─────────────────────────────────────────────────────────────

Uint8List _buildX3dhBindingBlob(List<int> x3dhPubSpki) {
  final prefix = utf8.encode(_x3dhBindingLabel);
  final out = Uint8List(prefix.length + x3dhPubSpki.length)
    ..setRange(0, prefix.length, prefix)
    ..setRange(prefix.length, prefix.length + x3dhPubSpki.length, x3dhPubSpki);
  return out;
}

/// Bundle of the long-term X3DH ECDH identity along with the ECDSA signature
/// that binds it to the primary identity.
class X3dhIdentity {
  const X3dhIdentity({
    required this.keyPair,
    required this.bindingSig,
    required this.pubSpki,
  });
  final EcKeyPair keyPair;
  final Uint8List bindingSig;
  final Uint8List pubSpki;
}

/// Return the cached long-term X3DH ECDH pair, creating+persisting one on
/// first call. The freshly-minted pub is signed with the ECDSA identity so
/// remote peers can bind the two halves together without a second TOFU step.
Future<X3dhIdentity> getOrCreateX3DHIdentity() async {
  if (_cachedX3dhKeyPair != null &&
      _cachedX3dhBindingSig != null &&
      _cachedX3dhPubSpki != null) {
    return X3dhIdentity(
      keyPair: _cachedX3dhKeyPair!,
      bindingSig: _cachedX3dhBindingSig!,
      pubSpki: _cachedX3dhPubSpki!,
    );
  }

  final existing = await keyStore().get(_keysTable, _x3dhKeyId);
  if (existing != null) {
    final pair = await _importKeyPair(existing, ecdsa: false);
    final pubSpki = Uint8List.fromList(existing['pubSpki'] as List<int>);
    final bindingSig =
        Uint8List.fromList(existing['bindingSig'] as List<int>);
    _cachedX3dhKeyPair = pair;
    _cachedX3dhPubSpki = pubSpki;
    _cachedX3dhBindingSig = bindingSig;
    return X3dhIdentity(
      keyPair: pair,
      bindingSig: bindingSig,
      pubSpki: pubSpki,
    );
  }

  final pair = await _ecdh.newKeyPair();
  final pubSpki = await _exportPubSpki(pair);
  final bindingSig = await signBytes(_buildX3dhBindingBlob(pubSpki));
  final row = await _serializeKeyPair(pair, pubSpki: pubSpki);
  await keyStore().put(_keysTable, {
    'id': _x3dhKeyId,
    ...row,
    'bindingSig': bindingSig,
    'createdAt': DateTime.now().millisecondsSinceEpoch,
  });
  _cachedX3dhKeyPair = pair;
  _cachedX3dhPubSpki = pubSpki;
  _cachedX3dhBindingSig = bindingSig;
  return X3dhIdentity(
    keyPair: pair,
    bindingSig: bindingSig,
    pubSpki: pubSpki,
  );
}

Future<Uint8List> exportX3DHIdentityPubSpki() async {
  if (_cachedX3dhPubSpki != null) return _cachedX3dhPubSpki!;
  final id = await getOrCreateX3DHIdentity();
  return id.pubSpki;
}

/// Verify an X3DH identity binding. Callers should also verify the ECDSA
/// identity itself (TOFU pin) — this only confirms the link.
Future<bool> verifyX3DHBinding(
  List<int> identitySpki,
  List<int> x3dhPubSpki,
  List<int> bindingSig,
) =>
    verifyWithRemoteSpki(
      identitySpki,
      _buildX3dhBindingBlob(x3dhPubSpki),
      bindingSig,
    );

/// Test helper — drop all in-memory caches so the next call re-reads storage.
void resetIdentityCacheForTests() {
  _cachedSigningKeyPair = null;
  _cachedSigningPubSpki = null;
  _cachedFingerprint = null;
  _cachedX3dhKeyPair = null;
  _cachedX3dhPubSpki = null;
  _cachedX3dhBindingSig = null;
}
